### ISSUE TITLE
[AArch64]: 128-bit Const Atomic Load implemented using Store Pair instruction, induces Runtime Crash on Arm AArch64

### ISSUE THREAD

**Issue by lukeg101** [llvm/llvm-project#61770]
Consider the following C test:
```
// test.c - fails when compiled and run with either LLVM or GCC
#include <stdatomic.h>
const _Atomic __int128 x = 0;
int main() {
__int128_t y = atomic_load_explicit(&x,memory_order_seq_cst);
}
// Fails when run as x is put into read-only section // $ clang -O0 test.c -o test.o
// $ ./test.o
/ / [1] 55891 bus error ./test.o
```
When compiled and run with clang (observed using apple clang on an M1 machine) targeting Arm AArch64, we observe a run-time crash.

As far as I can tell this fault is due to the fact that `const`-qualified data is stored in read-only section, and the compare-and-swap loop (https://godbolt.org/z/GaGKhT4f9) implemented using the `LDAXP`, `STLXP`, and `CBNZ` emits a store to the read-only memory.

It appears this issue is fixed by the patch [1] when targeting armv8.4-a and above, but it is not clear there is a (lock-free) fix for this crash for Armv8.3-a or below (or for any Arm arch that doesn't implement single-copy atomicity for 16-byte aligned 128-bit loads or stores). GCC emits a function call [2] instead. Is it possible to emit a function call in LLVM in a similar way without an ABI break?

[1] https://reviews.llvm.org/rG13aa102e07695297fd17f68913c343c95a7c56ad
[2] https://godbolt.org/z/7nqPW6Yb6

**Comment by Wilco1** [comment]
How can you use a const Atomic (since it can't ever change)? I think it makes more sense to worry about serious bugs such as the v8.4 sequence for SEQ_CST load being incorrect (a barrier is missing before the load, see https://gcc.gnu.org/bugzilla/show_bug.cgi?id=108891). Or the big-endian bug that incorrectly flips the low and high words in all 128-bit atomics: https://github.com/llvm/llvm-project/issues/61431

**Comment by lukeg101** [comment]
Hi Wilco, Thanks for this,
Indeed this const bug is an edge case of the language, but it still causes a compiler crash if anyone const casts a non-const atomic pointer. I believe that we should document this crash, even if we do not emit a warning when clang detects a const atomic load. It is not clear that there is a clear fix to this bug as we have discussed.

The endian bug I reported above has been triaged for fixing by the Arm compiler team.

As for the the compilation of the SC load, I generated some tests using Telechat and found a bug in LLVM that I have now reported upstream and mentioned to the compiler team :) https://github.com/llvm/llvm-project/issues/62652

**Comment by efriedma-quic** [comment]
It seems dubious to pessimize performance of atomics to handle this edge case.  The only atomic operation that even makes sense on a const global variable is a load, and that's useless on its own.  (The only possible use I can think of is some sort of sentinel value.)

Putting "const _Atomic __int128 x = 0;" into writable memory should be straightforward.

**Comment by lukeg101** [comment]
I agree we shouldn't change performance, but if possible we should document this behaviour at the very least, if not emit a sensible warning at least when we can statically check that an atomic load accesses a const pointer. 

Could you elaborate on what a sentinel value means in this context? I assume it means some kind of flag used to signify an observer has a lock on some data? This could make sense when considering Morello pointers?

**Comment by efriedma-quic** [comment]
By a "sentinel", I mean something like the last element of a linked list; not something you ever need to modify, but you'd mix it in with variables that are modifiable.

**Comment by gonzalobg** [comment]
AFAICT there are no APIs to concurrently modify the storage of an `const _Atomic`, `const std::atomic`, or `std::atomic_ref<const T>`. 

In the absence of concurrent modifications, non-atomic concurrent reads do not introduce data-races, so implementing the load using multiple non-atomic loads should be fine. 

That'd avoid regressing the performance of all programs using atomics to correctly implement this edge case.
